trieste/acquisition/rule.py

```diff
@@ -211,31 +212,208 @@ def acquire(
+
+        mask = tf.map_fn(is_not_in_points, self.pending_points, fn_output_signature=tf.bool)
+        return AsynchronousRuleState(tf.boolean_mask(self.pending_points, mask))
+
+    def add_points(self, new_points: TensorType) -> AsynchronousRuleState:
+        """Adds `new_points` to the already known pending points.
+
+        :param new_points: Points to add.
+        :return: New instance of `AsynchronousRuleState` with updated pending points.
+        """
+        if not self.has_pending_points:
+            return AsynchronousRuleState(new_points)
+
+        tf.debugging.assert_shapes(
+            [(self.pending_points, [..., "D"]), (new_points, [..., "D"])],
+            message=f"""Last dimension mismatch between pending points
+                        {tf.shape(self.pending_points)} and given points {tf.shape(new_points)}""",
+        )
+
+        new_pending_points = tf.concat([self.pending_points, new_points], axis=0)
+        return AsynchronousRuleState(new_pending_points)
+
+
+class AsynchronousOptimization(
+    AcquisitionRule[State[Optional["AsynchronousRuleState"], TensorType], SP_contra]
 ):
-    """AsynchronousGreedy rule, as name suggests,
-    is designed for asynchronous BO scenarios.
+    """AsynchronousOptimization rule is designed for asynchronous BO scenarios.
     By asynchronous BO we understand a use case when multiple objective function
     can be launched in parallel and are expected to arrive at different times.
     Instead of waiting for the rest of observations to return, we want to immediately
     use acquisition function to launch a new observation and avoid wasting computational resources.
 
     To make the best decision about next point to observe, acquisition function
     needs to be aware of currently running observations.
-    We call such points "pending". AsyncEGO provides a ``AsynchronousGreedy.State``
-    object that keeps track of pending points.
+    We call such points "pending", and consider them a part of acquisition state.
+    We use :class:`AsynchronousRuleState` to store these points.
+
+    AsynchronousOptimization works with non-greedy batch acquisition functions.
+    If there are P pending points, the acquisition function is used with batch size P+1.
+    During optimization first P points are fixed to pending, and thus optimization
+    is done over the last point only, which is then returned.
+
+    `Warning`: it is not clear how efficient this approach towards asynchronous BO is.
+    If you require efficiency, consider using :class:`AsynchronousGreedy`.
     """
 
-    @dataclass(frozen=True)
-    class State:
-        pending_points: TensorType
+    def __init__(
+        self,
+        builder: Optional[AcquisitionFunctionBuilder | SingleModelAcquisitionBuilder] = None,
+        optimizer: AcquisitionOptimizer[SP_contra] | None = None,
+    ):
+        """
+        :param builder: Batch acquisition function builder. Defaults to
```

Comment:
This code only works if the acquisition function can deal with batches. Not sure how clear this fact is from the docstring. Also, could we have a way to check this early?

Reply:
Docstring says "AsynchronousOptimization works with non-greedy batch acquisition functions." Shall we add something else?

For validation, we don't have a way in Trieste to validate if a function is a batch or non-batch. So the only way to do early validation is to compare against concrete classes. This is ugly and bound to break when we have more batch acquisitions, so for now we rely on acquisition function failing when called with more than one point.

Reply:
maybe something to think about in the future... maybe for now just add that it will fail with standard EI and work with MCEI?